code/game/objects/items/airlock_painter.dm

```diff
@@ -147,111 +150,309 @@
-	var/list/data = list()
-	data["decal_direction"] = stored_dir
-	data["decal_color"] = stored_color
-	data["decal_style"] = stored_decal
-	data["decal_list"] = list()
-	data["color_list"] = list()
-	data["dir_list"] = list()
-
-	for(var/i in decal_list)
-		data["decal_list"] += list(list(
-			"name" = i[1],
-			"decal" = i[2]
+/obj/item/airlock_painter/decal/ui_assets(mob/user)
+	. = ..()
+	. += get_asset_datum(spritesheet_type)
+
+/obj/item/airlock_painter/decal/ui_static_data(mob/user)
+	. = ..()
+	var/datum/asset/spritesheet/icon_assets = get_asset_datum(spritesheet_type)
+
+	.["icon_prefix"] = "[icon_assets.name]32x32"
+	.["supports_custom_color"] = supports_custom_color
+	.["decal_list"] = list()
+	.["color_list"] = list()
+	.["dir_list"] = list()
+	.["nondirectional_decals"] = nondirectional_decals
+
+	for(var/decal in decal_list)
+		.["decal_list"] += list(list(
+			"name" = decal[1],
+			"decal" = decal[2],
 		))
-	for(var/j in color_list)
-		data["color_list"] += list(list(
-			"colors" = j
+	for(var/color in color_list)
+		.["color_list"] += list(list(
+			"name" = color[1],
+			"color" = color[2],
 		))
-	for(var/k in dir_list)
-		data["dir_list"] += list(list(
-			"dirs" = k
+	for(var/dir in dir_list)
+		.["dir_list"] += list(list(
+			"name" = dir[1],
+			"dir" = dir[2],
 		))
-	return data
 
-/obj/item/airlock_painter/decal/ui_act(action,list/params)
+/obj/item/airlock_painter/decal/ui_data(mob/user)
+	. = ..()
+	.["current_decal"] = stored_decal
+	.["current_color"] = stored_color
+	.["current_dir"] = stored_dir
+	.["current_custom_color"] = stored_custom_color
+
+/obj/item/airlock_painter/decal/ui_act(action, list/params)
 	. = ..()
 	if(.)
 		return
 
 	switch(action)
 		//Lists of decals and designs
 		if("select decal")
-			var/selected_decal = params["decals"]
+			var/selected_decal = params["decal"]
+			var/selected_dir = text2num(params["dir"])
 			stored_decal = selected_decal
+			stored_dir = selected_dir
 		if("select color")
-			var/selected_color = params["colors"]
+			var/selected_color = params["color"]
 			stored_color = selected_color
-		if("selected direction")
-			var/selected_direction = text2num(params["dirs"])
-			stored_dir = selected_direction
+		if("pick custom color")
+			if(supports_custom_color)
+				var/chosen_color = input(usr, "Pick new color", "[src]", "yellow") as color|null
+				if(!chosen_color || QDELETED(src) || usr.incapacitated() || !usr.is_holding(src))
+					return
+				stored_custom_color = chosen_color
+				stored_color = chosen_color
 	update_decal_path()
 	. = TRUE
 
+/datum/asset/spritesheet/decals
+	name = "floor_decals"
+
+	/// The floor icon used for blend_preview_floor()
+	var/preview_floor_icon = 'icons/turf/floors.dmi'
+	/// The floor icon state used for blend_preview_floor()
+	var/preview_floor_state = "floor"
+	/// The associated decal painter type to grab decals, colors, etc from.
+	var/obj/item/airlock_painter/decal/painter_type = /obj/item/airlock_painter/decal
```

Comment:
A typepath isn't actually of the type /obj/item/airlock_painter/decal, so I don't think this variable should be typed. The only time you should use a typed variable and put something of a different type in it is when you are using initial() as there is no other way to do it.

Reply:
you're right, blame original PR maker lol